packages/cadl-ranch-specs/http/routes/main.tsp

```diff
@@ -0,0 +1,476 @@
+import "@typespec/http";
+import "@azure-tools/cadl-ranch-expect";
+
+using TypeSpec.Http;
+
+/**
+ * Define scenario in building the http route/uri
+ */
+@scenarioService("/routes")
+namespace Routes;
+
+@scenario
+@scenarioDoc("""
+  Simple operation at a fixed in an interface
+    Expected path: /routes/fixed
+  """)
+@route("fixed")
+op fixed(): void;
+
+@scenario
+@scenarioDoc("""
+  Simple operation at a fixed in an interface
+    Expected path: /routes/in-interface/fixed
+  """)
+@route("in-interface")
+interface InInterface {
+  @route("fixed")
+  fixed(): void;
+}
+
+@route("path")
+namespace PathParameters {
+  @scenario
+  @scenarioDoc("""
+    Path parameter defined implicitly
+      Value: "a"
+      Expected path: /routes/path/template-only/a
+    """)
+  @route("template-only/{param}")
+  op templateOnly(param: string): void;
+
+  @scenario
+  @scenarioDoc("""
+    Path parameter defined implicitly
+      Value: "a"
+      Expected path: /routes/path/explicit/a
+    """)
+  @route("explicit/{param}")
+  op explicit(@path param: string): void;
+
+  @scenario
+  @scenarioDoc("""
+    Path parameter annotated with @path but not defined explicitly in the route
+      Value: "a"
+      Expected path: /routes/path/annotation-only/a
+    """)
+  @route("annotation-only")
+  op annotationOnly(@path param: string): void;
+
+  @route("reserved-expansion")
+  namespace ReservedExpansion {
+    @scenario
+    @scenarioDoc("""
+      Defines a path parameter that shouldn't encode reserved characters. It should however still encode the other url characters.
+        Param value: "foo/bar baz"
+        Expected path: "/routes/path/reserved-expansion/template/foo/bar%20baz"
+      """)
+    @route("template/{+param}")
+    op template(param: string): void;
+
+    @scenario
+    @scenarioDoc("""
+      Defines a path parameter that shouldn't encode reserved characters. It should however still encode the other url characters.
+        Param value: "foo/bar baz"
+        Expected path: "/routes/path/reserved-expansion/annotation/foo/bar%20baz"
```

Comment:
What if user set param value `foo/bar%20baz`? Does it result in same path?

Reading the [RFC](https://datatracker.ietf.org/doc/html/rfc6570#section-3.2.3) but feeling unsure about such case.

Reply:
If the param is already encoded it shouldn't reencode I can add scenarios for that

Reply:
No need a new test. Just want to be sure of "what char still need encode on Reserved Expansion case".

Reply:
@timotheeguerin Could you help me understand the differences between the [x-ms-skip-url-encoding](https://github.com/Azure/autorest/blob/main/docs/extensions/readme.md#x-ms-skip-url-encoding) and [allowReserved](https://datatracker.ietf.org/doc/html/rfc6570#section-3.2.3) in uri template?

- x-ms-skip-url-encoding: Setting this as true would prevent the automatic encoding behavior for ALL characters
- allowReserved: Setting this as true would have RESERVED characters :/?#[]@!$&'()*+,;= in parameter values to be allowed to be sent as they are

My understanding is their scope for encoding or not is different. The former is for whole parameter value but latter is only working for reserved characters. If this is it, I would question on how to translate swagger extension to correct typespec.

```
path: template/{+param}

param value: foo/bar%20baz

x-ms-skip-url-encoding: template/foo/bar%20baz
allowReserved:  template/foo/bar%2520baz ==> pls note the % would also be pct-encoded as %25
```

Here is [one implementation](https://www.typescriptlang.org/play/?#code/GYVwdgxgLglg9mABAUwB4AcCGYAmBVAJxgBVkBbdAG0ymQApaLraAuRAZyiLAHMAaRADphAN0yUQydm2wBPANoBdAJRtO3HogDeAKET7ElZFERiJyAJK40iALyIADAG49BgsZAEkjKjWSD3Xwh6AHoAHS06eQBqAGJBEKcAfgAyRSTlOjokljCAdwAfAFJ5BwBaAE4AQTKAMUwy4EUtACYAXwL5QTCygD8AQgAqAHI6ZScWAAEU22iAEj4QpNiw+TDFFWjMnJY6MJxogrDB5WUksLaQ-kQ6MhoIAAsBOHRkAho4AgExIkwAIyMAjIcBwMGAMDeyjsAD5tK4DAYwTczJIrDgbND7CipIIjLwoA8oboESS3B4vIhhsMXKT9G14SSGQijCZsXZTOJJOx5Ni0WhotFFDTaUzEcBkZzkHZbPZwOjwWBkDgiaLSe4oJ4kFThaT6arEEjbiCwRCCCraSTDcDQeC3tL7MMRuaLbTDVUCO9ZIIYOx3Z66NjTnCXSGOeZ2djBAArOAwMB0F5vD4Ee2UkLDRBJNMZtjDPjDcb6i1tFCUdhSw1QWSvODitkyh1wP5R5DQAvB0MW+uIADyzdbUEEAGtkLJ2AHJcpBHd0HQhzDEAADOZaIdtWwr7HyIeKNqLqcxuN0YYpAs6zuIemdkvIMtS4kX-TdyNQOAAZS4cZ4Y0E7BAf3UL86AcAQsAIcsrCgI0bVNAQAEYHFOc9i31PULXYPIYCgR4bkTd5XzNDsQwgTBy0paJhhYItaXVTUUEgEFkAAJSkN4RCVCdzELTsSLI4ZYko6i1XJJBF1iFdkAY9EWPLAh2JwTjJGUPdkNpXipWGQRBMffRaIpRdBAkqTkDwJiLAAYTgCgEEkqDAxUoTEHU7MqJ0xA9NEkIjIgRiZLYjj7MXVTSWc6ltJ0jylycTdMF+AFkHXbzGNMiyrPQGywDsycHJ40iNKScLH0ixckhiuKjESrRJJ89EUss6zFUyxTkGUoLHNC09XIikSlxSMqYH+CqNyq4y6rSjKsq4nLQ3lTAQEoKAuqKnrquSsz6vSxrJqU4LLwZNpCz1UBIFgBB6Jq5jWLkgLJTUT9eFUDh7s0B9dJW0b1vGrbmqnQJqGCOgQnkfpRmUQZFCuAQIAXRcihXCBBEeWLLPRKooMQwRXw-DQ6DggA2KdXzwdBXgIcy8rGPdfuQIJQiKFpakh7NqdpwGigAZiqJnKILAIaf+umWnM7n80OnRxZCEJED866cEQABRDBsHYeAwB0HywHYOAjFxOBvzQLBcEIEhyF8Wg6AAIgeKAoHQFhJbQTAmH8HyyBCLRoh+S50SgTAYDLC2BAt4A4DgEIHlvSg4EQPJPkoHALaQxAdEl3sQFt9O2Gt237ZCR3ncRqyQhDsO-likIfb9stxY1rWdaj-WlaNohSCYPxLezu2HdQJ2qBdouPa9ivjCr9hA8QYPQ5CMuCHphwy4AL0T8Zk9Tnt0-QTPEE73P877wu3ZL8PI7gemAFYWgcWOCHj4fff99gdCAA) for uri template in JS and it do re-encode % twice. But I am not sure this is fully correct.

Reply:
The question is more why do you actually want to complexly skip encoding. Most of those cases are because they include / and it still make sense to encode uri elements.

If you look at JS functions it’s the difference between encodeUri(encode elements that are not valid in uri) and encodeUriComponent(encode the reserved chars)

Reply:
Let's case worst case of a customer passing an evil parameter (because if the parameter is well formed, it's doesn't really matter). If you skip all encoding, you likely will get a connection error (invalid url), or a 404 because the url doesn't make sense. If you escape special characters only, the url is always valid, but you will likely get a 404. In pragmatic terms, the different is not that big.
So I agree with Tim, that we could consider those equivalent, I dare you to find any cases where the user parameter would would have worked fine with no encoding, but would break if we encode some characters :)

Reply:
> ```
> allowReserved:  template/foo/bar%2520baz ==> pls note the % would also be pct-encoded as %25
> ```

Another online tester (using [rezigned/uri-template.js](https://github.com/rezigned/uri-template.js)) also apply this logic
![image](https://github.com/user-attachments/assets/c50c7af8-56b3-4ae5-bf9f-21e77dfedba4)